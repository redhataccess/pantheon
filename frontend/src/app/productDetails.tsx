import React, { Component } from 'react';
import {
    ActionGroup, Breadcrumb, BreadcrumbItem, Button, Form, FormGroup, Level, LevelItem, List, ListItem,
    Text, TextContent, TextVariants, TextInput
} from '@patternfly/react-core';

export interface IProps {
    productName: string
}

class ProductDetails extends Component<IProps, any> {
    public versionNames: string[] = [];

    constructor(props) {
        super(props);
        this.state = {
            allVersionNames: [],
            newVersion: ''
        };
    }

    public render() {
        return (
            <React.Fragment>
                {this.state.fetchProductDetails && this.fetchProductDetails(this.state.allVersionNames)}
                <div className="app-container">
                    <Breadcrumb>
                        <BreadcrumbItem>All Products</BreadcrumbItem>
                        <BreadcrumbItem>Product Details</BreadcrumbItem>
                    </Breadcrumb>
                </div>
                <div className="app-container">
                    <Level gutter="md">
                        <LevelItem>
                            <TextContent>
                                <Text component={TextVariants.h1}>{this.props.productName}{'  '}</Text>
                            </TextContent>
                        </LevelItem>
                        <LevelItem />
                    </Level>
                </div>
                {this.state.allVersionNames.length !== 0 && (
                    <div className="app-container">
                        Versions:
                    <List>
                            {this.state.allVersionNames.map((version) =>
                                <ListItem key={version}>{version}</ListItem>
                            )}
                        </List>
                    </div>)}
                <div className="app-container" />
                <div className="app-container">
                    <Form>
                        <div className="app-container">
                            <FormGroup
                                label="New Version:"
                                fieldId="new_version_name"
                            >
                                <TextInput id="new_version_name_text" type="text" placeholder="New version name" onChange={this.handleTextInputChange} value={this.state.newVersion} />
                            </FormGroup>
                            <ActionGroup>
                                <Button aria-label="Creates a new Version Name." onClick={this.saveVersion}>Save</Button>
                            </ActionGroup>
                        </div>
                    </Form>
                </div>
            </React.Fragment>
        );
    }

    private fetchProductDetails = (versionNames) => {
        // setup url fragment
        const urlFragment = this.props.productName.toString().toLowerCase().replace(/[^A-Z0-9]+/ig, "_");
        const path = '/content/products/' + urlFragment + '/versions.2.json'
        let key;
        versionNames = []

        fetch(path)
            .then((response) => {
                if (response.ok) {
                    return response.json();
                } else if (response.status === 404) {
                    // create versions path
                    this.createVersionsPath();
                    return versionNames;
                } else {
                    throw new Error(response.statusText);
                }
            })
            .then(responseJSON => {
                // tslint:disable-next-line: prefer-for-of
                for (let i = 0; i < Object.keys(responseJSON).length; i++) {
                    key = Object.keys(responseJSON)[i];
                    if ((key !== 'jcr:primaryType')) {
                        if (responseJSON[key].name !== undefined) {
                            versionNames.push(responseJSON[key].name);
                        }
                    }
                }
                this.setState({
                    allVersionNames: versionNames
                })
            })
            .catch((error) => {
                console.log(error)
            });
        return versionNames;
    };

    private handleTextInputChange = newVersion => {
        this.setState({ newVersion });
    };

    private saveVersion = () => {
        const formData = new FormData();
        formData.append("name", this.state.newVersion)
        formData.append("sling:resourceType", "pantheon/productVersion")
        formData.append("jcr:primaryType", 'pant:productVersion')

        const urlFragment = this.props.productName.toString().toLowerCase().replace(/[^A-Z0-9]+/ig, "_");
        const encodedVersion = this.state.newVersion.toString().toLowerCase().replace(/[^A-Z0-9]+/ig, "_");
        fetch(encodeURI('/content/products/' + urlFragment + '/versions/' + encodedVersion), {
            body: formData,
            method: 'post',
        }).then(response => {
            if (response.status === 200 || response.status === 201) {
                this.setState({ fetchProductDetails: true, newVersion: '' })
            } else {
                console.log('Version adding failure')
            }
        });

    };

    private createVersionsPath = () => {
        const formData = new FormData();
        const urlFragment = this.props.productName.toString().toLowerCase().replace(/[^A-Z0-9]+/ig, "_");
        fetch(encodeURI('/content/products/' + urlFragment + '/versions'), {
            body: formData,
            method: 'post',
        }).then(response => {
            if (response.status === 200 || response.status === 201) {
                console.log(" Created versions path " + response.status)
            } else {
                console.log(' Created versions path  failed!')

            }
        });
    }

}

export { ProductDetails }