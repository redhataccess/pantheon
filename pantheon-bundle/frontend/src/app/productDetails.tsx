import React, { useState, useEffect, useContext } from "react"
import {
    ActionGroup, Alert, AlertActionCloseButton, Breadcrumb, BreadcrumbItem, Button, Form, FormGroup, Level, LevelItem, List, ListItem,
    Text, TextContent, TextVariants, TextInput, InputGroup
} from "@patternfly/react-core"
import { Fields, JcrTypes, SlingTypes } from "@app/Constants"
import { ProductContext, IProduct } from "@app/contexts/ProductContext"


export const ProductDetails = (props: any) => {
    const allProducts = useContext(ProductContext);

    const [failedPost, setFailedPost] = useState(false)
    const [isMissingFields, setIsMissingFields] = useState(false)
    const [urlFragment, setUrlFragment] = useState("")
    const [isUrlFragmentValid, setIsUrlFragmentValid] = useState(true)
    const [newVersion, setNewVersion] = useState("")
    const [allVersionNames, setAllVersionNames] = useState([] as string[])
    const [currentProduct, setCurrentProduct] = useState(undefined as IProduct | undefined)

    useEffect(() => {
        if (allProducts) {
            const currentProduct = allProducts.find(a => a["jcr:uuid"] === props.match.params.id)
            setCurrentProduct(currentProduct)
            fetchProductVersions(currentProduct)
        }
    }, [allProducts])

    const fetchProductVersions = (currentProduct) => {
        const urlFragment = currentProduct && currentProduct.name.toLowerCase().replace(/[^A-Z0-9]+/ig, "_")
        const path = "/content/products/" + urlFragment + "/versions.2.json"
        let versionNames: string[] = []
        fetch(path).then((response) => {
            if (response.ok) {
                return response.json()
            } else if (response.status === 404) {
                // create versions path
                createVersionsPath(urlFragment)
                return versionNames
            } else {
                throw new Error(response.statusText)
            }
        }).then(json => {
            for (let i = 0; i < Object.keys(json).length; i++) {
                const key = Object.keys(json)[i]
                if (key !== Fields.JCR_PRIMARYTYPE && json[key].name !== undefined) {
                    versionNames.push(json[key].name)
                }
            }
            setAllVersionNames(versionNames)
        })
            .catch((error) => {
                console.log(error)
            })
    }

    const createVersionsPath = (urlFragment) => {
        const formData = new FormData()
        fetch(encodeURI("/content/products/" + urlFragment + "/versions"), {
            body: formData,
            method: "post",
        }).then(response => {
            if (response.status === 200 || response.status === 201) {
                console.log(" Created versions path " + response.status)
            } else {
                console.log(" Created versions path failed!")
            }
        })
    }

    const dismissNotification = () => {
        setFailedPost(false)
        setIsMissingFields(false)
    }

    const handleTextInputChange = newVersion => {
        setNewVersion(newVersion)
    }

    const handleUrlInputChange = urlFragment => {
        if (/^[-.\w]+$/.test(urlFragment)) {
            setUrlFragment(urlFragment)
            setIsUrlFragmentValid(true)
        } else {
            setIsUrlFragmentValid(false)
        }
    }

    const saveVersion = () => {
        if (newVersion === "" || urlFragment === "") {
            setIsMissingFields(true)
        } else {
            // creating formData obj and adding form's data to it
            const formData = new FormData()
            formData.append(Fields.NAME, newVersion)
            formData.append(Fields.SLING_RESOURCETYPE, SlingTypes.PRODUCT_VERSION)
            formData.append(Fields.JCR_PRIMARYTYPE, JcrTypes.PRODUCT_VERSION)
            formData.append(Fields.URL_FRAGMENT, urlFragment)

            // creating url-friendly version of product name and product version
            const productUrlFragment = currentProduct && currentProduct.name.toLowerCase().replace(/[^A-Z0-9]+/ig, "_")
            const encodedVersion = newVersion.toString().toLowerCase().replace(/[^A-Z0-9]+/ig, "_")
            // sending formData to backend
            fetch(encodeURI("/content/products/" + productUrlFragment + "/versions/" + encodedVersion), {
                body: formData,
                method: "post",
            }).then(response => {
                if (response.status === 200 || response.status === 201) {
                    setNewVersion("")
                    setUrlFragment("")
                    currentProduct && fetchProductVersions(currentProduct)
                } else {
                    setFailedPost(true)
                    setIsMissingFields(false)
                    console.log("Version adding failure")
                }
            })
        }
    }

    return (
        <React.Fragment>
            <div className="app-container">
                <Breadcrumb>
                    <BreadcrumbItem ><a href="#/products">All Products</a></BreadcrumbItem>
                    <BreadcrumbItem to="#" isActive={true}>Product Details</BreadcrumbItem>
                </Breadcrumb>
            </div>
            {currentProduct && <div className="app-container">
                <Level>
                    <LevelItem>
                        <TextContent>
                            <Text component={TextVariants.h1}>{currentProduct.name}{"  "}</Text>
                        </TextContent>
                    </LevelItem>
                    <LevelItem />
                </Level>
            </div>}
            {allVersionNames.length !== 0 && (
                <div className="app-container">
                    Versions:
                    <List>
                        {allVersionNames.map((version) =>
                            <ListItem key={version}>{version}</ListItem>
                        )}
                    </List>
                </div>)}
            <br />
            <div>
                <Form>
                    <div className="pant-form-width-md">
                        {isMissingFields &&
                            <div className="notification-container">
                                <Alert variant="warning"
                                    title="Fields indicated by * are mandatory"
                                    actionClose={<AlertActionCloseButton onClose={dismissNotification} />}
                                />
                            </div>
                        }
                        {!isUrlFragmentValid &&
                            <div className="notification-container">
                                <Alert variant="warning"
                                    title="Allowed input for ulrFragment: alphanumeric, hyphen, period and underscore"
                                    actionClose={<AlertActionCloseButton onClose={dismissNotification} />}
                                />
                            </div>
                        }
                        <FormGroup
                            label="New Version:"
                            isRequired={true}
                            fieldId="new_version_name"
                        >
                            <TextInput id="new_version_name_text" type="text" placeholder="New version name" onChange={handleTextInputChange} value={newVersion} />
                        </FormGroup>
                        <br />
                        <FormGroup
                            label=" URL Fragment:"
                            isRequired={true}
                            fieldId="url_fragment"
                        >
                            <TextInput id="url_fragment_text" type="text" placeholder="Url fragment" onChange={handleUrlInputChange} value={urlFragment} />
                        </FormGroup>
                        {failedPost &&
                            <div className="notification-container">
                                <Alert
                                    variant="danger"
                                    title="Failed to create product version."
                                    actionClose={<AlertActionCloseButton onClose={dismissNotification} />}
                                >
                                    Please check if you are logged in as a publisher.
                                    </Alert>
                            </div>
                        }
                        {allProducts && allProducts.length > 0 &&
                            <ActionGroup>
                                <Button aria-label="Creates a new Version Name." onClick={saveVersion}>Save</Button>
                            </ActionGroup>
                        }
                    </div>
                </Form>
            </div>
        </React.Fragment>
    )
}
