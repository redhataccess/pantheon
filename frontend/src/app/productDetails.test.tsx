import React from 'react';
import { ProductDetails } from '@app/productDetails';
import "isomorphic-fetch"

import { mount, shallow } from 'enzyme';
import { Breadcrumb, Button, Form, FormGroup, Level, LevelItem, TextContent, TextInput, Text } from '@patternfly/react-core';
import renderer from 'react-test-renderer';
import sinon from 'sinon'

const props = {
  productName: "Red Hat Enterprise Linux"
}

describe('ProductDetails tests', () => {
  test('should render ProductDetails component', () => {
    const view = shallow(<ProductDetails productName="Red Hat Enterprise Linux" />);
    expect(view).toMatchSnapshot();
  });

  it('should render a form', () => {
    const wrapper = mount(<ProductDetails productName="Red Hat Enterprise Linux" />);
    const form = wrapper.find(Form);
    expect(form.exists()).toBe(true)
  });

  it('should render a form group', () => {
    const wrapper = mount(<ProductDetails productName="Red Hat Enterprise Linux" />);
    const formGroup = wrapper.find(FormGroup);
    expect(formGroup.exists()).toBe(true)
  });

  it('should render a text input', () => {
    const wrapper = mount(<ProductDetails productName="Red Hat Enterprise Linux" />);
    const textInput = wrapper.find(TextInput);
    expect(textInput.exists()).toBe(true)
  });

  it('should render a Button', () => {
    const wrapper = mount(<ProductDetails productName="Red Hat Enterprise Linux" />);
    const button = wrapper.find(Button);
    expect(button.exists()).toBe(true)
  });

  it('should render a breadcrumb', () => {
    const wrapper = mount(<ProductDetails productName="Red Hat Enterprise Linux" />);
    const breadCrumb = wrapper.find(Breadcrumb);
    expect(breadCrumb.exists()).toBe(true)
  });

  it('should render a Level element', () => {
    const wrapper = mount(<ProductDetails productName="Red Hat Enterprise Linux" />);
    const level = wrapper.find(Level);
    expect(level.exists()).toBe(true)
  });

  it('should render a LevelItem', () => {
    const wrapper = mount(<ProductDetails productName="Red Hat Enterprise Linux" />);
    const levelItem = wrapper.find(LevelItem);
    expect(levelItem.exists()).toBe(true)
  });

  it('should render a TextContent element', () => {
    const wrapper = mount(<ProductDetails productName="Red Hat Enterprise Linux" />);
    const textContent = wrapper.find(TextContent);
    expect(textContent.exists()).toBe(true)
  });

  it('should render a Text element', () => {
    const wrapper = mount(<ProductDetails productName="Red Hat Enterprise Linux" />);
    const text = wrapper.find(Text);
    expect(text.exists()).toBe(true)
  });

  it('test props', () => {

    const productDetails = mount(<ProductDetails {...props} />).text
    expect(productDetails.length === 1)
  });

  it('test click event', () => {
    const mockCallBack = jest.fn();

    const button = shallow((<Button onClick={mockCallBack}>Submit</Button>));
    button.find('button').simulate('click');
    expect(mockCallBack.mock.calls.length).toEqual(1);
  });

  it('test fetchProductDetails function', () => {
    const wrapper = renderer.create(<ProductDetails {...props} />);
    const inst = wrapper.getInstance();
    expect(inst.fetchProductDetails([])).toMatchSnapshot();
  });

  it('test handleTextInputChange function', () => {
    const wrapper = renderer.create(<ProductDetails {...props} />);
    const inst = wrapper.getInstance();
    expect(inst.handleTextInputChange("1.1")).toMatchSnapshot();
  });

  it('test saveVersion function', () => {
    const wrapper = renderer.create(<ProductDetails {...props} />);
    const inst = wrapper.getInstance();
    expect(inst.saveVersion).toMatchSnapshot();
  });

  it('test createVersionsPath function', () => {
    const wrapper = renderer.create(<ProductDetails {...props} />);
    const inst = wrapper.getInstance();
    expect(inst.createVersionsPath).toMatchSnapshot();
  });

  test('newSearch() click event', () => {
    const wrapper = shallow(<ProductDetails {...props} />);
    const instance = wrapper.instance();
    const spy = sinon.spy(instance, 'saveVersion');

    wrapper.setState({ "newVersion": "1.1" })
    wrapper.find(Button).simulate('click');
    sinon.assert.called(spy);
  });

  it('test componentWillReceiveProps function', () => {
    const wrapper = renderer.create(<ProductDetails {...props} />);
    const inst = wrapper.getInstance();
    expect(inst.componentWillReceiveProps({...props})).toMatchSnapshot();
  });
});