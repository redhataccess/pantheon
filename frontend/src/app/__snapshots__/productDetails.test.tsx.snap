// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`ProductDetails tests should render ProductDetails component 1`] = `
<Fragment>
  <div
    className="app-container"
  >
    <Breadcrumb>
      <BreadcrumbItem>
        <a
          href="#/products"
          onClick={[Function]}
        >
          All Products
        </a>
      </BreadcrumbItem>
      <BreadcrumbItem
        isActive={true}
        to="#"
      >
        Product Details
      </BreadcrumbItem>
    </Breadcrumb>
  </div>
  <div
    className="app-container"
  >
    <Level
      gutter="md"
    >
      <LevelItem>
        <TextContent>
          <Text
            component="h1"
          >
            Red Hat Enterprise Linux
              
          </Text>
        </TextContent>
      </LevelItem>
      <LevelItem />
    </Level>
  </div>
  <div
    className="app-container"
  />
  <div
    className="app-container"
  >
    <Form>
      <div
        className="app-container"
      >
        <FormGroup
          fieldId="new_version_name"
          label="New Version:"
        >
          <TextInput
            aria-label={null}
            className=""
            id="new_version_name_text"
            isDisabled={false}
            isReadOnly={false}
            isRequired={false}
            isValid={true}
            onChange={[Function]}
            placeholder="New version name"
            type="text"
            value=""
          />
        </FormGroup>
        <ActionGroup>
          <Button
            aria-label="Creates a new Version Name."
            onClick={[Function]}
          >
            Save
          </Button>
        </ActionGroup>
      </div>
    </Form>
  </div>
</Fragment>
`;

exports[`ProductDetails tests test componentWillReceiveProps function 1`] = `undefined`;

exports[`ProductDetails tests test createVersionsPath function 1`] = `[Function]`;

exports[`ProductDetails tests test fetchProductDetails function 1`] = `Array []`;

exports[`ProductDetails tests test handleTextInputChange function 1`] = `undefined`;

exports[`ProductDetails tests test saveVersion function 1`] = `[Function]`;
